add new get-kubeconfig action to kubernetes-master

xiwenc opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Once the k8s cluster is completed the config with login credentials is stored at `/home/ubuntu/config`. With this juju action we enable users to pull out credentials to connect to the cluster. 

Alternatively, we could use `juju scp --proxy=true` but python libjuju (https://pythonlibjuju.readthedocs.io/en/latest/_modules/juju/machine.html#Machine.scp_from) doesn't support *proxy* yet as we require it.


**Which issue(s) this PR fixes**:
NONE

**Special notes for your reviewer**:
Usecase has been discussed with @tvansteenburgh 

**Does this PR introduce a user-facing change?**:
```release-note
Add new get-kubeconfig action to kubernetes-master
```

tvansteenburgh commented on cluster/juju/layers/kubernetes-master/actions/get-kubeconfig:
```diff
@@ -0,0 +1,35 @@
+    Read the kubeconfig on the master and return it as JSON
+    """
+    kubeconfig = load(_kubectl(["config", "view", "-o", "yaml", "--raw"]))
+    action_set({"credential": kubeconfig})
```
I would prefer this be "kubeconfig" instead of "credential" since that more explicitly describes the output value.

xiwenc replied:
Good point! Will change that.

tvansteenburgh commented on cluster/juju/layers/kubernetes-master/actions/get-kubeconfig:
```diff
@@ -0,0 +1,35 @@
+
+def get_kubeconfig():
+    """
+    Read the kubeconfig on the master and return it as JSON
```
@xiwenc Can you remind me why you want JSON instead of yaml? It might be nice to have a parameter so the user can choose which format they want.

Before I realized that you were converting to JSON, I was wondering if this whole action could be replaced with:
```
#!/bin/bash
action-set "kubeconfig" $(cat /home/ubuntu/config)
```

xiwenc replied:
The choice for JSON was the assumption that JSON format is being used in the action response. Other than that we don't have a preference.

The bash oneliner looks tempting but we prefer to use Python with more control on format checking and error handling.

Cynerva commented on cluster/juju/layers/kubernetes-master/actions/get-kubeconfig:
```diff
@@ -0,0 +1,35 @@
+    Read the kubeconfig on the master and return it as JSON
+    """
+    kubeconfig = load(_kubectl(["config", "view", "-o", "yaml", "--raw"]))
+    action_set({"credential": kubeconfig})
```
Passing the loaded data directly to `action_set` results in messy action output:

```
$ juju run-action kubernetes-master/0 get-kubeconfig --wait
unit-kubernetes-master-0:
  id: 9846184a-8766-495a-87f2-29ecd7ba18ad
  results:
    credential: '{''apiVersion'': ''v1'', ''clusters'': [{''cluster'': {''certificate-authority-data'':
      ''REDACTED'',
      ''server'': ''https://REDACTED:6443''}, ''name'': ''juju-cluster''}], ''contexts'':
      [{''context'': {''cluster'': ''juju-cluster'', ''user'': ''admin''}, ''name'':
      ''juju-context''}], ''current-context'': ''juju-context'', ''kind'': ''Config'',
      ''preferences'': {}, ''users'': [{''name'': ''admin'', ''user'': {''password'':
      ''REDACTED'', ''username'': ''admin''}}]}'
  status: completed
  timing:
    completed: 2019-03-19 21:56:08 +0000 UTC
    enqueued: 2019-03-19 21:56:07 +0000 UTC
    started: 2019-03-19 21:56:08 +0000 UTC
  unit: kubernetes-master/0
```

You can see the action output presenting the kubeconfig here, but it's a string that containers neither YAML nor JSON. It's a Python formatted string, which you can see the `action_set` function doing [here](https://github.com/juju/charm-helpers/blob/master/charmhelpers/core/hookenv.py#L964).

I recommend skipping the `yaml.load` and instead passing the raw kubeconfig string to `action_set`.

If it really needs to be JSON, then keep the `yaml.load`, but use `json.dumps` before you pass it to `action_set`.

xiwenc replied:
Thanks for the extensive test and great recommendations! Will rewrite this action and submit for review again.

xiwenc commented:
Thanks all for the feedback! Will update this PR shortly!

xiwenc commented:
I've updated the PR. The output looks a lot cleaner now:

```
ubuntu@juju-d40a41-0:~$ juju run-action kubernetes-master/0 get-kubeconfig --wait
unit-kubernetes-master-0:
  id: 7e3b3f87-5681-4b53-8471-bf67f42e59cb
  results:
    kubeconfig: '{"kind": "Config", "apiVersion": "v1", "preferences": {}, "clusters":
      [{"name": "juju-cluster", "cluster": {"server": "https://172.16.20.72:6443",
      "certificate-authority-data": "REDACTED"}}],
      "users": [{"name": "admin", "user": {"username": "admin", "password": "REDACTED"}}],
      "contexts": [{"name": "juju-context", "context": {"cluster": "juju-cluster",
      "user": "admin"}}], "current-context": "juju-context"}'
  status: completed
  timing:
    completed: 2019-03-25 12:29:03 +0000 UTC
    enqueued: 2019-03-25 12:29:00 +0000 UTC
    started: 2019-03-25 12:29:02 +0000 UTC
  unit: kubernetes-master/0
```

Summary of changes:
- JSON format verification
- renamed *credentials* to *kubeconfig*
- return JSON string of kubeconfig as the value of kubeconfig due to how `action_set` [behaves](https://github.com/juju/charm-helpers/blob/master/charmhelpers/core/hookenv.py#L964) as [pointed out](https://github.com/juju-solutions/kubernetes/pull/221#discussion_r267125213).

Cynerva commented:
LGTM, thanks!
